zcash_primitives/src/note_encryption.rs

```diff
@@ -29,6 +64,58 @@ impl Default for Memo {
     }
 }
 
+impl PartialEq for Memo {
+    fn eq(&self, rhs: &Memo) -> bool {
+        &self.0[..] == &rhs.0[..]
+    }
+}
+
+impl Memo {
+    /// Returns a Memo containing the given slice, appending with zero bytes if necessary,
+    /// or None if the slice is too long. If the slice is empty, Memo::default() is
+    /// returned.
+    pub fn from_bytes(memo: &[u8]) -> Option<Memo> {
+        if memo.is_empty() {
+            Some(Memo::default())
+        } else if memo.len() <= 512 {
+            let mut data = [0; 512];
+            data[0..memo.len()].copy_from_slice(memo);
+            Some(Memo(data))
+        } else {
+            // memo is too long
+            None
+        }
+    }
+
+    /// Returns a Memo containing the given string, or None if the string is too long.
+    pub fn from_str(memo: &str) -> Option<Memo> {
+        Memo::from_bytes(memo.as_bytes())
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..]
+    }
+
+    /// Returns:
+    /// - None if the memo is not text
+    /// - Some(Ok(memo)) if the memo contains a valid UTF8 string
+    /// - Some(Err(e)) if the memo contains invalid UTF8
+    pub fn to_utf8(&self) -> Option<Result<String, str::Utf8Error>> {
```

Comment:
I think having Option<Result<...>> is confusing as Option and Result exist to serve similar purposes. I could see this being justified if we handles Err() and None differently, but in this code it doesn't look like we are doing so.
This function could be changed to:
```
pub fn to_utf8(&self) -> Option<String> {
```
if we changed the line (112):
```
Some(str::from_utf8(data).map(|memo| memo.to_owned()))
```
to
```
match str::from_utf8(data).map(|memo| memo.to_owned()) {
    Ok(utf8) => Some(utf8),
    Err(_) => None,
}
```

Reply:
The reason I separated the two cases was so that callers of this API _could_ handle the `Err()` differently from the `None`. See [here](https://github.com/str4d/librustzcash/blob/ae63116f639847e54716fbf47f19589d12038a16/zcash_client_sqlite/src/lib.rs#L544-L551) as an example (where I end up inverting the `Result` and `Option` so that I can combine the error types due to SQL errors vs the memo being invalid).

Given that the error in question here is specifically that the memo field contained invalid UTF-8 when it was supposed to be UTF-8 text, one alternative is to just do as you are, and pretend the memo does not exist if it is invalid. Another option would be to use `String::from_utf8_lossy` instead, and try to return as much useful text as we can. The current case (return the error) gives more information to the caller (who could then choose to manually call `String::from_utf8_lossy` on the memo's bytes if they wanted), but does forgo a slightly simpler interface.

Reply:
I am fine with leaving this as is.